src/utils-lgpl/ffmpeg_compress_plugin.c

```diff
@@ -0,0 +1,547 @@
+#define COMPR_PLAYBACK_MAX_NUM_FRAGMENTS (8)
+
+struct ffmpeg_compress_priv {
+	struct snd_compr_caps compr_cap;
+	struct pcm *pcm;
+	unsigned int out_rate;
+	unsigned int out_channels;
+	enum pcm_format out_format;
+
+	AVPacket *pkt;
+	AVCodec *codec;
+	AVCodecParserContext *parser;
+	AVCodecContext *context;
+	AVFrame *decoded_frame;
+};
+
+enum AVCodecID ffmpeg_get_avcodec_id(uint32_t codec_id)
+{
+	switch(codec_id) {
+	case SND_AUDIOCODEC_FLAC:
+		return AV_CODEC_ID_FLAC;
+	case SND_AUDIOCODEC_MP3:
+		return AV_CODEC_ID_MP3;
+	default:
+		break;
+	}
+	return AV_CODEC_ID_NONE;
+}
+
+static enum AVSampleFormat pcm_format_to_avformat(enum pcm_format format)
+{
+	switch (format) {
+	case PCM_FORMAT_S8:
+		return AV_SAMPLE_FMT_U8;
+	case PCM_FORMAT_S16_LE:
+	case PCM_FORMAT_S16_BE:
+		return AV_SAMPLE_FMT_S16;
+
+	case PCM_FORMAT_S24_LE:
+	case PCM_FORMAT_S24_BE:
+	case PCM_FORMAT_S24_3LE:
+	case PCM_FORMAT_S24_3BE:
+	case PCM_FORMAT_S32_LE:
+	case PCM_FORMAT_S32_BE:
+		return AV_SAMPLE_FMT_S32;
+	default:
+		return AV_SAMPLE_FMT_NONE;
+	};
+	return AV_SAMPLE_FMT_NONE;
+}
+
+int decode_frame_from_packet(struct ffmpeg_compress_priv *priv, AVFrame *decoded_frame)
+{
+	int16_t *samples;
+	int64_t src_ch_layout, dst_ch_layout;
+	int src_rate, dst_rate;
+	uint8_t **src_data = NULL, **dst_data = NULL;
+	int src_nb_channels = 0, dst_nb_channels = 0;
+	int src_linesize, dst_linesize;
+	int src_nb_samples, dst_nb_samples;
+	enum AVSampleFormat src_sample_fmt, dst_sample_fmt;
+	int dst_bufsize;
+	struct SwrContext *swr_ctx = NULL;
+	int ret;
+	AVCodecContext* codec = priv->context;
+
+	if (decoded_frame->channel_layout == 0)
+		decoded_frame->channel_layout = av_get_default_channel_layout(decoded_frame->channels);
+
+	src_rate = decoded_frame->sample_rate;
+	dst_rate = priv->out_rate;
+	src_ch_layout = decoded_frame->channel_layout;
+	dst_ch_layout = av_get_default_channel_layout(priv->out_channels);
+	src_sample_fmt = decoded_frame->format;
+	dst_sample_fmt = pcm_format_to_avformat(priv->out_format);
+
+	if ((src_sample_fmt != dst_sample_fmt) || (src_rate != dst_rate) ||
+	    (src_ch_layout != dst_ch_layout)) {
+		src_nb_samples = decoded_frame->nb_samples;
+		src_linesize = (int) decoded_frame->linesize;
+		src_data = decoded_frame->data;
+
+		swr_ctx = swr_alloc();
+		if (!swr_ctx) {
+			fprintf(stderr, "Could not allocate resampler context\n");
+			ret = -ENOMEM;
+			goto exit;
+		}
+
+		av_opt_set_int(swr_ctx, "in_channel_layout", src_ch_layout, 0);
+		av_opt_set_int(swr_ctx, "out_channel_layout", dst_ch_layout,  0);
+		av_opt_set_int(swr_ctx, "in_sample_rate", src_rate, 0);
+		av_opt_set_int(swr_ctx, "out_sample_rate", dst_rate, 0);
+		av_opt_set_sample_fmt(swr_ctx, "in_sample_fmt", src_sample_fmt, 0);
+		av_opt_set_sample_fmt(swr_ctx, "out_sample_fmt", dst_sample_fmt,  0);
+
+		/* initialize the resampling context */
+		if ((ret = swr_init(swr_ctx)) < 0) {
+			fprintf(stderr, "Failed to initialize the resampling context\n");
```

Comment:
Reference taken from https://github.com/FFmpeg/FFmpeg/blob/master/doc/examples/decode_audio.c and  https://github.com/FFmpeg/FFmpeg/blob/master/doc/examples/resampling_audio.c